File: SiriusQuality/SiriusQuality-Priestly-Taylor/src/apsim/SiriusQuality_Priestly_Taylor/PotentialET_PTComponent.cs
using Models.Core;
using Models.Utilities;
using System; 
namespace Models.Crop2ML;
     

/// <summary>
///  PotentialET_PT component
/// </summary>
public class PotentialET_PTComponent 
{

    /// <summary>
    ///  constructor of PotentialET_PT component
    /// </summary>
    public PotentialET_PTComponent() {}

    //Declaration of the associated strategies
    PriestlyTaylor _PriestlyTaylor = new PriestlyTaylor();

    /// <summary>
    /// Gets and sets the latent heat of vaporization of water
    /// </summary>
    [Description("latent heat of vaporization of water")] 
    [Units("MJ kg-1")] 
    public double lambdaV
    {
        get
        {
             return _PriestlyTaylor.lambdaV; 
        }
        set
        {
            _PriestlyTaylor.lambdaV = value;
        }
    }

    /// <summary>
    /// Gets and sets the psychrometric constant
    /// </summary>
    [Description("psychrometric constant")] 
    [Units("")] 
    public double psychrometricConstant
    {
        get
        {
             return _PriestlyTaylor.psychrometricConstant; 
        }
        set
        {
            _PriestlyTaylor.psychrometricConstant = value;
        }
    }

    /// <summary>
    /// Gets and sets the Priestley-Taylor evapotranspiration proportionality constant
    /// </summary>
    [Description("Priestley-Taylor evapotranspiration proportionality constant")] 
    [Units("")] 
    public double Alpha
    {
        get
        {
             return _PriestlyTaylor.Alpha; 
        }
        set
        {
            _PriestlyTaylor.Alpha = value;
        }
    }

    /// <summary>
    /// Gets and sets the hour of the day if the component is hourly, -999 if the component is daily
    /// </summary>
    [Description("hour of the day if the component is hourly, -999 if the component is daily")] 
    [Units("")] 
    public int ih
    {
        get
        {
             return _PriestlyTaylor.ih; 
        }
        set
        {
            _PriestlyTaylor.ih = value;
        }
    }

    /// <summary>
    /// Algorithm of PotentialET_PT component
    /// </summary>
    public void CalculateModel(PotentialET_PTState s,PotentialET_PTState s1,PotentialET_PTRate r,PotentialET_PTAuxiliary a,PotentialET_PTExogenous ex)
    {
        _PriestlyTaylor.CalculateModel(s,s1, r, a, ex);
    }

    /// <summary>
    /// Initialization of PotentialET_PT component
    /// </summary>
    public void Init(PotentialET_PTState s, PotentialET_PTState s1, PotentialET_PTRate r, PotentialET_PTAuxiliary a, PotentialET_PTExogenous ex)
    {
    }

    /// <summary>
    /// constructor copy of PotentialET_PT component
    /// </summary>
    /// <param name="toCopy"></param>
    public PotentialET_PTComponent(PotentialET_PTComponent toCopy): this() // copy constructor 
    {
        lambdaV = toCopy.lambdaV;
        psychrometricConstant = toCopy.psychrometricConstant;
        Alpha = toCopy.Alpha;
        ih = toCopy.ih;
}
}

File: SiriusQuality/SiriusQuality-Priestly-Taylor/src/dssat/SiriusQuality_Priestly_Taylor/PotentialET_PTComponent.f90
MODULE Potentialet_ptmod
    USE Priestlytaylormod
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_potentialet_pt(lambdaV, &
        psychrometricConstant, &
        Alpha, &
        solarRadiation, &
        hslope, &
        ih, &
        netRadiation, &
        evapoTranspirationPriestlyTaylor)
        IMPLICIT NONE
        INTEGER:: i_cyml_r
        REAL, INTENT(IN) :: lambdaV
        REAL, INTENT(IN) :: psychrometricConstant
        REAL, INTENT(IN) :: Alpha
        REAL, INTENT(IN) :: solarRadiation
        REAL, INTENT(IN) :: hslope
        INTEGER, INTENT(IN) :: ih
        REAL, INTENT(IN) :: netRadiation
        REAL, INTENT(OUT) :: evapoTranspirationPriestlyTaylor
        !- Name: PotentialET_PT -Version: 1.0, -Time step: 1
        !- Description:
    !            * Title: PotentialET Component
    !            * Authors: SQ
    !            * Reference: None
    !            * Institution: INRAE
    !            * ExtendedDescription: https://pimlday26.sciencesconf.org/program?lang=en
    !            * ShortDescription: Potential ET (based on Penman and Priestly-Taylor)
        !- inputs:
    !            * name: lambdaV
    !                          ** description : latent heat of vaporization of water
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 10.0
    !                          ** min : 0.0
    !                          ** default : 2.454
    !                          ** unit : MJ kg-1
    !            * name: psychrometricConstant
    !                          ** description : psychrometric constant
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 1.0
    !                          ** min : 0.0
    !                          ** default : 0.66
    !                          ** unit : 
    !            * name: Alpha
    !                          ** description : Priestley-Taylor evapotranspiration proportionality constant
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 100.0
    !                          ** min : 0.0
    !                          ** default : 1.5
    !                          ** unit : 
    !            * name: solarRadiation
    !                          ** description : solar Radiation
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 1000.0
    !                          ** min : 0.0
    !                          ** default : 3.0
    !                          ** unit : MJ m-2 d-1
    !            * name: hslope
    !                          ** description : the slope of saturated vapor pressure temperature curve at a given temperature
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 1000.0
    !                          ** min : 0.0
    !                          ** default : 0.584
    !                          ** unit : hPa degC-1
    !            * name: ih
    !                          ** description : hour of the day if the component is hourly, -999 if the component is daily
    !                          ** inputtype : variable
    !                          ** parametercategory : constant
    !                          ** datatype : INT
    !                          ** max : 24
    !                          ** min : -999
    !                          ** default : -999
    !                          ** unit : 
    !            * name: netRadiation
    !                          ** description : net radiation
    !                          ** inputtype : variable
    !                          ** variablecategory : auxiliary
    !                          ** datatype : DOUBLE
    !                          ** max : 5000.0
    !                          ** min : 0.0
    !                          ** default : 1.566
    !                          ** unit : MJ m-2 d-1
        !- outputs:
    !            * name: evapoTranspirationPriestlyTaylor
    !                          ** description : evapoTranspiration of Priestly Taylor
    !                          ** datatype : DOUBLE
    !                          ** variablecategory : rate
    !                          ** max : 10000.0
    !                          ** min : 0.0
    !                          ** unit : g m-2 d-1
        call model_priestlytaylor(lambdaV, netRadiation,  &
                psychrometricConstant, Alpha, solarRadiation, hslope,  &
                ih,evapoTranspirationPriestlyTaylor)
    END SUBROUTINE model_potentialet_pt

END MODULE
